aspnet/2-structured-data/Models/DatastoreBookStore.cs

```diff
@@ -132,112 +129,81 @@ public DatastoreBookStore(string projectId)
-            var commitRequest = new CommitRequest()
-            {
-                Mode = "NON_TRANSACTIONAL",
-                Mutation = mutation
-            };
-            return _datastore.Datasets.Commit(commitRequest, _projectId)
-                .Execute();
+            _datastore = DatastoreClient.Create();
+            // I like this better.
+            _foo = DatastoreFoo.Create(_projectId);
         }
-        // [END commitmutation]
 
         // [START create]
         public void Create(Book book)
         {
-            var result = CommitMutation(new Mutation()
-            {
-                InsertAutoId = new Entity[] { book.ToEntity() }
-            });
-            book.Id = result.MutationResult.InsertAutoIdKeys.First().Path.First().Id.Value;
+            // P0 Calling .ToInsert() is very weird.
+            // Having two very different ways to insert, update, etc. depending on whether or not
+            // I'm in a transaction is annoying.  I want one *interface* to do it.
+            // How about a NullTransaction where all operations are immediately committed, and the
+            // final .Commit() is a no-op?  Or, make DatastoreFoo and Transaction implement the
+            // same interface?
+            CommitResponse response = _datastore.Commit(_projectId, Mode.NonTransactional, new[] { book.ToEntity().ToInsert() });
+            Key key = response.MutationResults[0].Key;
+            book.Id = key.Path.First().Id;
         }
         // [END create]
 
         public void Delete(long id)
         {
-            CommitMutation(new Mutation()
-            {
-                Delete = new Key[] { id.ToKey() }
-            });
+            // Pretty good.
+            _foo.Delete(id.ToKey());
+            var trans = _foo.BeginTransaction();
+            trans.Delete(id.ToKey());
+            trans.Commit();
         }
 
         // [START list]
         public BookList List(int pageSize, string nextPageToken)
         {
             var query = new Query()
             {
+                // Why do see a squiggly red line under "Book", but not in your snippet?
+                Kind = { "Book" },
                 Limit = pageSize,
-                Kinds = new[] { new KindExpression() { Name = "Book" } },
             };
 
             if (!string.IsNullOrWhiteSpace(nextPageToken))
                 query.StartCursor = nextPageToken;
 
-            var datastoreRequest = _datastore.Datasets.RunQuery(
-                datasetId: _projectId,
-                body: new RunQueryRequest() { Query = query }
-            );
+            var response = _foo.RunQuery(query);
 
-            var response = datastoreRequest.Execute();
             var results = response.Batch.EntityResults;
             var books = results.Select(result => result.Entity.ToBook());
 
             return new BookList()
             {
                 Books = books,
+                // More string vs proto byte string warnings below.  Why?

```

Comment:
Because Datastore uses ByteString as its page token. With `DatastoreFoo` we expose a page-streamed query form and you can extract the cursor as a string, which is basically the base64-converted form of the bytestring.